public/app/core/utils/EventsProvider.tsx

```diff
@@ -0,0 +1,101 @@
+import React, { ComponentType, FunctionComponent, useEffect } from 'react';
+import { Observable, Subject, Unsubscribable } from 'rxjs';
+import { filter } from 'rxjs/operators';
+import { AppEvent, Subtract } from '@grafana/data';
+
+import appEvents from '../app_events';
+
+export interface EventsContextApi<SubscribeEvents, PublishEvents> {
+  events: (originFilter: string) => Observable<AppEvent<SubscribeEvents>>;
+  publish: (event: AppEvent<PublishEvents>, origin: string, payload?: PublishEvents) => void;
+}
+
+export interface RootEventsContextApi {
+  rootEvents: Observable<AppEvent<any>>;
+  rootPublish: (event: AppEvent<any>, payload?: any) => void;
+}
+
+const events = new Subject<AppEvent<any>>();
+const publish = <T extends {} = {}>(event: AppEvent<T>, payload?: T) => {
+  events.next({ ...event, payload });
+  appEvents.emit(event, payload);
+};
+
+export const rootEventing: RootEventsContextApi = {
+  rootEvents: events,
+  rootPublish: publish,
+};
+
+export const EventsContext = React.createContext(rootEventing);
+EventsContext.displayName = 'EventsContext';
+
+export const EventsProvider = ({ children }: { children: React.ReactNode }) => {
+  return <EventsContext.Provider value={rootEventing}>{children}</EventsContext.Provider>;
+};
+
+export const provideEvents = (component: ComponentType<any>) => (props: any) => (
+  <EventsProvider>{React.createElement(component, { ...props })}</EventsProvider>
+);
+
+export const withEvents = <SubscribeEvents extends {}, PublishEvents extends {}>(
+  ...subscribeEvents: Array<AppEvent<SubscribeEvents>>
+) => {
+  type ComponentWithEventingType = EventsContextApi<SubscribeEvents, PublishEvents>;
+  return <C extends ComponentWithEventingType>(Component: ComponentType<C>) => {
+    type ComponentWithOutEventingType = Subtract<C, ComponentWithEventingType>;
+
+    const ComponentWithEventing: FunctionComponent<ComponentWithOutEventingType> = (props: any) => {
+      useEffect(() => {
+        return function unsubscribe() {
+          if (subscription) {
+            subscription.unsubscribe();
+            subscription = null;
+          }
+          events.unsubscribe();
+        };
+      }, []);
+      let subscription: Unsubscribable | null = null;
+      const events = new Subject<AppEvent<SubscribeEvents>>();
+      const subscribe = (origin: string): Observable<AppEvent<SubscribeEvents>> => {
```

Comment:
Seems weird that you need to specify `origin` here as it is probably not a good idea to change it during the lifetime of the component. It also needs to match publishing call so seems a bit error prone.

Reply:
The `origin` here should be renamed it's actually a filter so that the component doesn't receive it's own events.